### PR TITLE
TSK-1674: Added received timestamp to task

### PR DESCRIPTION
**Pull request by Tristan2357**
https://sonarcloud.io/dashboard?branch=TSK-1674&id=Tristan2357_taskana

---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 1]
```diff
@@ -44,6 +47,7 @@
   @BeforeEach
   @WithAccessId(user = "admin")
   void setUp() throws Exception {
+    resetDb(false);
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:50]: why?

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 2]
```diff
@@ -85,6 +89,52 @@ void should_UpdateTaskCorrectlyInDatabase_When_AddingAnAttachment() throws Excep
...
 
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_UpdateTaskReceived_When_AddingAnAttachment() throws Exception {
+    task.addAttachment(attachment);
+    task = taskService.updateTask(task);
+    task = taskService.getTask(task.getId());
+    assertThat(task)
+        .extracting(TaskSummary::getReceived)
+        .isEqualTo(LocalDate.parse("2018-01-15").atStartOfDay().toInstant(ZoneOffset.UTC));
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:100]: Instant.parse?

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 3]
```diff
@@ -85,6 +89,52 @@ void should_UpdateTaskCorrectlyInDatabase_When_AddingAnAttachment() throws Excep
...
+  }
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_NotUpdateTaskReceived_When_TaskAlreadyHasAReceived() throws Exception {
+    task.setReceived(Instant.parse("2019-09-13T08:44:17.588Z"));
+    task.addAttachment(attachment);
+    task = taskService.updateTask(task);
+    task = taskService.getTask(task.getId());
+    assertThat(task).extracting(TaskSummary::getReceived).isNotEqualTo(attachment.getReceived());
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:110]:
```suggestion
    assertThat(task.getReceived()).isNotEqualTo(attachment.getReceived());
```

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAccTest.java [hunk 4]
```diff
@@ -324,4 +325,27 @@ void should_UpdateCallbackInfo_When_RequestedByApi() throws Exception {
...
+    Task createdTask = taskService.createTask(newTask);
+
+    Task retrievedTask = taskService.getTask(createdTask.getId());
+    Instant retrievedTime = Instant.parse("2019-09-13T08:44:17.588Z");
+    retrievedTask.setReceived(retrievedTime);
+    taskService.updateTask(retrievedTask);
+
+    Task retrievedUpdatedTask = taskService.getTask(createdTask.getId());
+
+    assertThat(retrievedUpdatedTask).extracting(TaskSummary::getReceived).isEqualTo(retrievedTime);
```
[the remark this answers is not in the document]
Comment (mustaphazorgati) [reply lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAccTest.java:347]: Maybe something for the COP?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 5]
```diff
@@ -565,6 +565,44 @@ void should_NotGetEmptyAttachmentList_When_GettingTaskWithAttachment() {
...
+    String url =
+        restHelper.toUrl(RestEndpoints.URL_TASKS_ID, "TKI:000000000000000000000000000000000024");
+    HttpEntity<Object> auth = new HttpEntity<>(RestHelper.generateHeadersForUser("admin"));
+
+    ResponseEntity<TaskRepresentationModel> response =
+        TEMPLATE.exchange(url, HttpMethod.GET, auth, TASK_MODEL_TYPE);
+
+    TaskRepresentationModel repModel = response.getBody();
+    assertThat(repModel).isNotNull();
+    assertThat(repModel).extracting(TaskSummaryRepresentationModel::getReceived).isNotNull();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:579]: combine assertions :)

File: web/src/app/workplace/components/task-status-details/task-status-details.component.html [hunk 6]
```diff
@@ -66,10 +78,10 @@
 
     <!-- TRANSFERRED -->
```
Comment (mustaphazorgati) [inline web/src/app/workplace/components/task-status-details/task-status-details.component.html:79]: comment doesn't match

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]
:clap:

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskQuerySqlProvider.java [hunk 7]
```diff
@@ -319,6 +237,44 @@ public static String queryTaskColumnValues() {
         + CLOSING_SCRIPT_TAG;
   }
 
+  private static String commonSelectFields() {
+    // includes only the names that start with a t, because other columns are conditional
+    return Arrays.stream(TaskQueryColumnName.values())
+        .map(TaskQueryColumnName::toString)
+        .filter(column -> column.startsWith("t"))
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskQuerySqlProvider.java:244]: i think this would not improve readability, as i it would include negation and would still need a cast to string

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskQuerySqlProvider.java [hunk 8]
```diff
@@ -319,6 +237,44 @@ public static String queryTaskColumnValues() {
         + CLOSING_SCRIPT_TAG;
   }
 
+  private static String commonSelectFields() {
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskQuerySqlProvider.java:240]: this could be improved, but i think this should be done in a COP + a continuation of TSK-1290

File: lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java [hunk 9]
```diff
@@ -276,16 +285,14 @@ void testCreateExternalTaskWithAttachment() throws Exception {
...
-          .contains("\"Property_7\":\"Property Value of Property_7\"")
-          .contains("\"Property_6\":\"Property Value of Property_6\"");
+      Set<String> expectedPhrasesSet =
+          IntStream.rangeClosed(1, 27)
+              .mapToObj(Integer::toString)
+              .map(
+                  number ->
+                      "\"Property_" + number + "\":\"Property Value of Property_" + number + "\"")
+              .collect(Collectors.toSet());
+      assertThat(customProperties).contains(expectedPhrasesSet);
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java:295]: as customProperties is a String, it does not have a containsExactly
but a rewrite could be done, if desired

File: lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java [hunk 10]
```diff
@@ -349,7 +356,7 @@ void testCreateExternalTaskWithMultipleAttachments() throws Exception {
                 "ArchiveId",
                 "12345678901234567890123456789012345678901234567890"),
             "E-MAIL",
-            "2018-01-15",
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java:352]: to test a sideeffect, which was in the specification of the task

> If a attachment with a received information exists and the task does not have any received information, the received timestamp is propagated from the attachment to the task.

we assumed it should use the earlier received timestamp, which requires an earlier and a later received date, to test it

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAccTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAccTest.java:347]: > General Rule: Do not use any operation within the assertThat() function. Use existing functions from AssertJ’s assert classes.

> We always use method references whenever possible

Both statements are found in our backend coding guidelines
i think my code adheres to them, whereas yours does not

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java (hunk 1 above)
Comment (Tristan2357) [reply lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:50]: our tests previously did not reset the db, so the introduction of a new test led to the failure of other tests, when executed sequentially
this line resets the state of the db
it seems costly, but leads to more consistent tests

the need for this line shows, that it could be time to rethink at least parts of our testing strategy

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 5 above)
Comment (Tristan2357) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:579]: actually combine, or just drop the first?

File: lib/taskana-core/src/test/java/acceptance/task/QueryTasksAccTest.java [hunk 11]
```diff
@@ -462,6 +462,23 @@ void testQueryAllPaged() {
...
+        TASK_SERVICE
+            .createTaskQuery()
+            .receivedWithin(new TimeInterval(null, Instant.parse("2018-01-29T15:55:20Z")))
+            .list();
+    long resultCount =
+        TASK_SERVICE
+            .createTaskQuery()
+            .receivedWithin(new TimeInterval(null, Instant.parse("2018-01-29T15:55:18Z")))
+            .count();
+    assertThat(results).hasSize(22);
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply lib/taskana-core/src/test/java/acceptance/task/QueryTasksAccTest.java:478]: i changed it to identical timestamps
and will add more tests for these scenarios

File: common/taskana-common-data/src/main/resources/sql/monitor-data/monitor-sample-data.sql [hunk 12]
```diff
@@ -28,54 +28,54 @@ INSERT INTO ATTACHMENT VALUES('ATT:000000000000000000000000000000000009', 'TKI:0
...
+INSERT INTO TASK VALUES('TKI:000000000000000000000000000000000001', 'ETI:000000000000000000000000000000000001', RELATIVE_DATE(0) , RELATIVE_DATE(0) , RELATIVE_DATE(0) , RELATIVE_DATE(0) , RELATIVE_DATE(-5), null    , RELATIVE_DATE(-70000), 'Task01',   'teamlead-1',   'Some description.', 'Some custom Note', 1       , 'READY'  , 'EXTERN'                , 'L10000'          , 'CLI:000000000000000000000000000000000001', 'WBI:000000000000000000000000000000000001', 'USER-1-1'    , 'DOMAIN_C', 'BPI21'            , 'PBPI21'                  , 'John', 'MyCompany1', 'MySystem1', 'MyInstance1', 'MyType1', 'MyValue1', true   , false         , null         , 'NONE'        , null             , 'Geschaeftsstelle A'   , 'Vollkasko'           , null    , null    , null    , null    , null    , null    , null    , null     , null     , null     , null     , null     , null     , 'VALUE_01'      );
```
Comment (holgerhagen) [inline common/taskana-common-data/src/main/resources/sql/monitor-data/monitor-sample-data.sql:32]: as an example: this statement would fail if the database was updated with one of our update scripts. (see other comment on the sequence).

**Review by holgerhagen — CHANGES_REQUESTED** [review]

File: common/taskana-common/src/main/resources/sql/db2/taskana-schema-db2.sql [hunk 13]
```diff
@@ -68,6 +68,7 @@ CREATE TABLE TASK (
         COMPLETED TIMESTAMP NULL,
         MODIFIED TIMESTAMP NULL,
         PLANNED TIMESTAMP NULL,
+        RECEIVED TIMESTAMP NULL,
```
[the remark this answers is not in the document]
Comment (Tristan2357) [reply common/taskana-common/src/main/resources/sql/db2/taskana-schema-db2.sql:71]: the update script and the schema generate slightly different tables, but if we follow best practices for sql this should not lead to any problems
i can update our sample-data scripts, but this would lead to lower readability and maintainability in my opinion
other solutions could be reordering the table in the update script which would lead to a more complex update script, which would be harder to understand
or i could move the required field to the end of the table, which would disturb the order of columns in my opinion

i would argue, that this does not actually need a fix, as our sample data is only being used by us with our schema and all other database expressions are generated in a way, that ignores actual column order

**Review by mustaphazorgati — APPROVED** [review]

**Review by holgerhagen — APPROVED** [review]
